### PR TITLE
Add sink/source pull streams

### PR DESCRIPTION
**Pull request by dy**

### CODE REVIEW

File: source.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+var infinite = require('pull-stream/sources/infinite');
+var createReader = require('./reader');
+
+module.exports = source;
+
+
+function source (node, options) {
+	let read = createReader(node, options);
+
+	//FIXME: too clumsy. I’d expect pull.asyncSource(read)
```
Comment (dy) [inline source.js]: @ahdinosaur is there a more graceful solution for async source? (see below)
[pull-async](http://npmjs.org/package/pull-async) is nice but seems to be called once.

Comment (dy) [reply source.js]: Alright, I just took modified _pull/sources/infinite_ code to be async.

File: sink.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+function sink (node, options) {
+	let write = createWriter(node, options);
+
+	let stream = pull(
+		asyncMap(write),
+		drain()
+	);
+
+	stream.abort = write.end;
```
Comment (ahdinosaur) [inline sink.js]: should this call the drain stream's `.abort`, or does this work as is? i have a feeling it wouldn't end the stream till the next time it's trying to write.

File: test.js [hunk 3]
```diff
@@ -186,6 +190,51 @@ test('Readable', function (done) {
...
+
+	pull(
+		source,
+		// pull.take(10),
+		sink
+	);
+
+	setTimeout(() => {
+		sink.abort();
+		//FIXME: how to end stream externally?
```
Comment (ahdinosaur) [inline test.js]: maybe want to test that the source and sink have actually ended?

**Comment by ahdinosaur** [comment]
nice work on this @dfcreative, looks really great. :smile:

File: sink.js (hunk 2 above)
Comment (dy) [reply sink.js]: yeah, it will not end stream, it will just stop invoking callbacks. Should fix.
Drain has abort?

File: source.js (hunk 1 above)
Comment (ahdinosaur) [reply source.js]: nice! i agree there should be something more graceful, _pull-async_ was meant for a specific use case (turning a single continuable into a stream, kinda like an async version of `pull.once`), keen for a re-usable module that does what you're doing.

File: sink.js (hunk 2 above)
Comment (ahdinosaur) [reply sink.js]: yep, the return stream from `pull.drain()` has a `.abort` function.

File: test.js (hunk 3 above)
Comment (dy) [reply test.js]: how can we test that?

**Pull request merged**